pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -1195,9 +1196,10 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 	for _, pInfo := range podInfoList {
 		// Since there may be many gated pods and they will not move from the
 		// unschedulable pool, we skip calling the expensive isPodWorthRequeueing.
-		if pInfo.Gated {
+		if pInfo.Gated && pInfo.UnschedulablePlugins.Has(names.SchedulingGates) {
```

Comment:
Seems we're target the optimization for scheduling gates plugin, can we make this more clear like checking the pod.spec.schedulingGates.

Reply:
I prefer the current version; I wouldn't like to make any part of scheduling queue depend on the implementation detail of the scheduling gate plugin.

Reply:
Is it accurate that we're considering the pod `Gated` if [any of the plugins return unschedulable](https://github.com/kubernetes/kubernetes/blob/ef9965ebc66dafda37800bb04f5e284535bbba10/pkg/scheduler/internal/queue/scheduling_queue.go#L560)? I suppose that it should be true IFF the SchedulingGates plugin returns unschedulable.

Reply:
The scheduling queue considers a Pod to be `Gated` if any PreEnqueue returns Unschedulable, yes. And, there could be multiple PreEnqueue plugins; actually we have two PreEnqueue plugins (when the DRA feature gate is enabled)

Reply:
I understand that that is the current behavior, but my question was more: is that the desired future state? From its name, `Gated` seems to refer to `Pod.Spec.SchedulingGates` being set. But the concept of Gated (at least in scheduling queue) is now mixed with any `PreEnqueuePlugin` returning unschedulable. I see how this became to be the state - as the first and only (at the time) `PreEnqueuePlugin` was the `SchedulingGates` plugin - added in the same [commit](https://github.com/kubernetes/kubernetes/pull/113275/commits/0b27f25252d827afe2cac2d9ab67595275315608) which defined the logic of `pInfo.Gated = !p.runPreEnqueuePlugins(...)`

Reply:
> is that the desired future state?

I agree that `Gated` comes from scheduling gates and may be confusing (For this point, we can rename the field if necessary), but I think the current way of implementing PreEnqueue logic in the scheduling queue is OK with it.

Reply:
So, whether is your point; we should rename it, or we should use `Gated` only for the scheduling gate plugin and do something another for other PreEnqueue plugins? (or anything else?)

Reply:
> In this PR, however, we are introducing a potentially expensive operation (set membership) into a performance critical section, rather than just reading a boolean.

So, (leaving the naming issue aside) is your core concern in this PR having `pInfo.UnschedulablePlugins.Has` in this shortcut path from a performance perspective? 

I'm not sure if we could see a visible performance difference between `pInfo.UnschedulablePlugins.Has` (= O(1) map access [internally](https://github.com/kubernetes/kubernetes/blob/fb7bbd2e5df36212ca5133bcbaa46dcdb4d18f0e/staging/src/k8s.io/apimachinery/pkg/util/sets/set.go#L78)) vs `pInfo.SchedulingGated` (= new field to cache a result of schedulinggate plugin).

With a test case https://github.com/kubernetes/kubernetes/pull/125550, I checked the throughput in each case and didn't see a difference actually. (You can try on your machine too.)

<details><summary>results</summary>

**Revert https://github.com/kubernetes/kubernetes/pull/124618**

(In order to ensure the test case could reproduce the degradation you reported in https://github.com/kubernetes/kubernetes/issues/124384.)

```
    {
      "data": {
        "Average": 220.9635550403882,
        "Perc50": 214.9945877262486,
        "Perc90": 262.0829689158846,
        "Perc95": 262.0829689158846,
        "Perc99": 262.0829689158846
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 2371.50060887,
        "Perc50": 2345.0626566416036,
        "Perc90": 4528.77598152425,
        "Perc95": 4824.387990762125,
        "Perc99": 5060.877598152425
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_pod_scheduling_duration_seconds",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

**Current master** 

sample1:

```
    {
      "data": {
        "Average": 525.6099658687274,
        "Perc50": 525.6099658687274,
        "Perc90": 525.6099658687274,
        "Perc95": 525.6099658687274,
        "Perc99": 525.6099658687274
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 884.6911246259998,
        "Perc50": 904.0471512770139,
        "Perc90": 1923.18407960199,
        "Perc95": 2241.5920398009953,
        "Perc99": 2496.318407960199
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_pod_scheduling_duration_seconds",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

sample2:

```
    {
      "data": {
        "Average": 571.9668219205608,
        "Perc50": 571.9668219205608,
        "Perc90": 571.9668219205608,
        "Perc95": 571.9668219205608,
        "Perc99": 571.9668219205608
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 961.9552155919993,
        "Perc50": 1011.3989637305698,
        "Perc90": 2061.9455252918287,
        "Perc95": 2310.9727626459144,
        "Perc99": 2510.1945525291826
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_pod_scheduling_duration_seconds",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

**This PR**

sample1:

```
    {
      "data": {
        "Average": 629.98727425706,
        "Perc50": 629.98727425706,
        "Perc90": 629.98727425706,
        "Perc95": 629.98727425706,
        "Perc99": 629.98727425706
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 838.6084798769984,
        "Perc50": 849.5971978984238,
        "Perc90": 1456.5517241379312,
        "Perc95": 2008.2758620689658,
        "Perc99": 2449.6551724137935
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_pod_scheduling_duration_seconds",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    }
```

sample2:

```
    {
      "data": {
        "Average": 481.7783482353274,
        "Perc50": 481.7783482353274,
        "Perc90": 481.7783482353274,
        "Perc95": 481.7783482353274,
        "Perc99": 481.7783482353274
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 831.9787021770011,
        "Perc50": 832.9241877256318,
        "Perc90": 1427.2566371681417,
        "Perc95": 1993.628318584071,
        "Perc99": 2446.7256637168143
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_pod_scheduling_duration_seconds",
        "Name": "SchedulingGatedPodsImpactsThroughput/5000Nodes/namespace-2",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```


</details>

Reply:
Just reran tests with https://github.com/kubernetes/kubernetes/pull/125504, and got similar results to the ones you posted. Units are average pods/s

base run1: 166.7
base run2: 167.1

#125527 run1: 153.9
#125527 run2: 156.4

Reply:
Results from a new version of https://github.com/kubernetes/kubernetes/pull/125550:

Revert #124618: 59 pods/s (avg) / 12ms scheduler_scheduling_algorithm_duration_seconds (avg)
master: 149 pods/s (avg) / 4.19 ms scheduler_scheduling_algorithm_duration_seconds (avg)
This PR: 145 pods/s (avg) / 4ms scheduler_scheduling_algorithm_duration_seconds (avg)

Anyways, I think we can say this PR (= `pInfo.UnschedulablePlugins.Has`) wouldn't make any visible degradation from the master branch.

Reply:
updated the comment.